pkg/security/security_profile/profile/manager.go

```diff
@@ -511,62 +537,100 @@ func (m *SecurityProfileManager) unlinkProfile(profile *SecurityProfile, workloa
 		return
 	}
 
-	// if time.Now()-event.ContainerContext.CreatedAt < time.Second*30 {
-	// 	// TODO: put the event in a cache to be pop back after x sec to have a chance to
-	// 	// retrieve a profile for that workload
-	// }
-
 	selector, err := cgroupModel.NewWorkloadSelector(utils.GetTagValue("image_name", event.ContainerContext.Tags), utils.GetTagValue("image_tag", event.ContainerContext.Tags))
 	if err != nil {
 		return
 	}
+
+	// lookup profile
 	profile := m.GetProfile(selector)
 	if profile == nil || profile.Status == 0 {
-		m.eventFilteringNoProfile[evtType].Inc()
+		m.eventFiltering[event.GetEventType()][NoProfile].Inc()
 		return
 	}
 
 	FillProfileContextFromProfile(&event.SecurityProfileContext, profile)
 
-	processNodes := profile.findProfileProcessNodes(event.ProcessContext)
-	if len(processNodes) == 0 {
-		m.eventFilteringAbsent[evtType].Inc()
-		return
+	// check if the event should be injected in the profile automatically
+	if profile.autolearnEnabled {
+		if autoLearned, autoLearnErr := m.tryAutolearn(profile, event); autoLearnErr != nil || autoLearned {
+			return
+		}
 	}
 
-	switch evtType {
-	// for fork/exec/exit events, as we already found some nodes, no need to investigate further
-	case model.ExecEventType:
+	// check if the event is in its profile
+	ok, err := profile.ActivityTree.Contains(event, activity_tree.SecurityProfile)
+	if err != nil {
+		// ignore, evaluation failed
+		m.eventFiltering[event.GetEventType()][NoProfile].Inc()
+		return
+	}
+	if ok {
 		event.AddToFlags(model.EventFlagsSecurityProfileInProfile)
+		m.eventFiltering[event.GetEventType()][InProfile].Inc()
+	} else {
+		m.eventFiltering[event.GetEventType()][NotInProfile].Inc()
+	}
+}
 
-	case model.DNSEventType:
-		if findDNSInNodes(processNodes, event) {
-			event.AddToFlags(model.EventFlagsSecurityProfileInProfile)
-		}
+// tryAutolearn tries to autolearn the input event. Returns true if the event was autolearned.
+func (m *SecurityProfileManager) tryAutolearn(profile *SecurityProfile, event *model.Event) (bool, error) {
+	// have we reached the stable state time limit ?
+	if profile.lastAnomalyNano == 0 {
+		profile.lastAnomalyNano = event.TimestampRaw
+	}
+	if time.Duration(event.TimestampRaw-profile.lastAnomalyNano) >= m.config.RuntimeSecurity.AnomalyDetectionMinimumStableDelayBeforeAnomalyDetection {
+		profile.autolearnEnabled = false
+		return false, nil
+	}
+
+	// have we reached the unstable time limit ?
+	if time.Duration(event.TimestampRaw-profile.loadedNano) >= m.config.RuntimeSecurity.AnomalyDetectionUnstableProfileTimeThreshold {
+		m.eventFiltering[event.GetEventType()][UnstableProfile].Inc()
+		return false, fmt.Errorf("unstable profile: time limit reached")
+	}
+
+	// check if the unstable size limit was reached
+	if profile.ActivityTree.Stats.ApproximateSize() >= m.config.RuntimeSecurity.AnomalyDetectionUnstableProfileSizeThreshold {
+		m.eventFiltering[event.GetEventType()][UnstableProfile].Inc()
```

Comment:
should we add `profile.autolearnEnabled = false` when we reach the safe guards ? (here and on the previous if)